### PR TITLE
Implement offsetParent/Top/Left/Width/Height.

### PR DESCRIPTION
**Pull request by glennw**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6784)

### CODE REVIEW

**Comment by glennw** [comment]
This PR fixes a large number of layout issues in the enyojs sampler. But I'm not sure if the implementation could be made a lot simpler, it feels a bit complicated. Comments?

r? @jdm for script bits and @pcwalton for the layout bits?

File: components/layout/layout_task.rs [hunk 1]
```diff
@@ -127,6 +127,9 @@ pub struct LayoutTaskData {
     /// A queued response for the content boxes of a node.
     pub content_boxes_response: Vec<Rect<Au>>,
 
+    /// A queued response for the offset parent/rect of a node.
+    pub offset_parent_response: OffsetParentResponse,
+
```
Comment (pcwalton) [inline components/layout/layout_task.rs]: This is probably going to get unwieldy over time. Perhaps we should factor this out into a struct now?

File: components/layout/flow.rs [hunk 2]
```diff
@@ -284,6 +284,7 @@ pub trait Flow: fmt::Debug + Sync {
     /// Iterates through border boxes of all of this flow's fragments.
     fn iterate_through_fragment_border_boxes(&self,
                                              iterator: &mut FragmentBorderBoxIterator,
+                                             level: i32,
```
Comment (pcwalton) [inline components/layout/flow.rs]: Add a comment documenting the level parameter.

File: components/layout/layout_task.rs [hunk 3]
```diff
@@ -1362,7 +1399,7 @@ impl UnioningFragmentBorderBoxIterator {
 }
 
 impl FragmentBorderBoxIterator for UnioningFragmentBorderBoxIterator {
-    fn process(&mut self, _: &Fragment, border_box: &Rect<Au>) {
+    fn process(&mut self, _: &Fragment, _: i32,border_box: &Rect<Au>) {
```
Comment (pcwalton) [inline components/layout/layout_task.rs]: uber-nit: space after `i32,` for consistency

File: components/script/dom/window.rs [hunk 4]
```diff
@@ -769,6 +770,28 @@ impl<'a> WindowHelpers for &'a Window {
         rects
     }
 
+    fn offset_parent_query(self, node: TrustedNodeAddress) -> (Option<Root<Element>>, Rect<Au>) {
+        self.reflow(ReflowGoal::ForScriptQuery,
+                    ReflowQueryType::OffsetParentQuery(node),
+                    ReflowReason::Query);
+        self.join_layout(); //FIXME: is this necessary, or is layout_rpc's mutex good enough?
```
Comment (pcwalton) [inline components/script/dom/window.rs]: I think it's not necessary

**Comment by glennw** [comment]
Ref https://github.com/servo/servo/issues/6643

File: components/layout/layout_task.rs (hunk 1 above)
Comment (glennw) [inline components/layout/layout_task.rs]: @jdm opened https://github.com/servo/servo/issues/6787 - perhaps we can do this as a follow up?

Comment (jdm) [inline components/layout/layout_task.rs]: It is, in fact, prime E-easy material :)

**Comment by glennw** [comment]
@pcwalton Addressed those comments and rebased.

**Comment by metajack** [comment]
```
/css21_dev/html4/first-letter-dynamic-003a.htm
----------------------------------------------
FAIL [Parent]
/css21_dev/html4/first-letter-dynamic-003b.htm
----------------------------------------------
FAIL [Parent]
/css21_dev/html4/run-in-basic-017.htm
-------------------------------------
FAIL [Parent]
/css21_dev/html4/run-in-relpos-between-003.htm
----------------------------------------------
CRASH expected FAIL [Parent]
```

**Comment by glennw** [comment]
@pcwalton re-r? Fixed a panic in the css tests and updated expectations. If you're fine with that change, I'll squash and rebase after that.

**Comment by jdm** [comment]
@bors-servo: r-
http://mxr.mozilla.org/servo/source/tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/input-type-button.html?force=1#35
That test is legitimately passing now. Please look at failures that occur?

**Comment by glennw** [comment]
@jdm Yep, I just managed to repro it locally, so will look into it.

**Comment by glennw** [comment]
@jdm Ah, I actually misread that build log - there was a subtest in input-type-button that was previously failing, and this PR fixes that subtest. I have updated the last commit to remove that test failure from the WPT metadata.

**Comment by jdm** [comment]
That ini file can be removed now, if it contains no other failures.

**Comment by glennw** [comment]
Updated to remove that ini file.

**Pull request merged**